llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp

```diff
@@ -1244,17 +1247,45 @@ InstructionCost SystemZTTIImpl::getInterleavedMemoryOpCost(
   return NumVectorMemOps + NumPermutes;
 }
 
-static int getVectorIntrinsicInstrCost(Intrinsic::ID ID, Type *RetTy) {
+static int getVectorIntrinsicInstrCost(
+    Intrinsic::ID ID,
+    Type *RetTy,
+    const SmallVectorImpl<Type *>& ParamTys) {
   if (RetTy->isVectorTy() && ID == Intrinsic::bswap)
     return getNumVectorRegs(RetTy); // VPERM
+  else if (ID == Intrinsic::vector_reduce_add) {
+    // Retrieve number and size of elements for the vector op
+    auto *VTy = cast<FixedVectorType>(ParamTys.front());
+    unsigned NumElements = VTy->getNumElements();
+    unsigned BitWidth = VTy->getScalarSizeInBits();
+    // For scalar sizes <128 bits, we fall back to the generic cost estimate
+    if (BitWidth > SystemZ::VectorBits)
+      return -1;
+    // How many elements can a single vector register hold
+    unsigned MaxElemsPerVector = SystemZ::VectorBits / BitWidth;
+    // How many vector regs are needed to represent the input elements (V)
+    unsigned VectorRegsNeeded = getNumVectorRegs(VTy);
+    // How many instructions are needed for the final sum of vector elems (S)
+    unsigned LastVectorHandling = std::min(NumElements,
+      2 * (unsigned) std::round(std::log2(MaxElemsPerVector)));
+    // We need V load operations to read the input. Then, we use vector adds to
+    // create a sum vector, which takes V/2 + V/4 + ... = V - 1 operations.
+    // Finally, we need S operations to sum up the elements of the sum vector,
+    // and a final store operation, for a total of 2 * V + S operations.
+    // We'll ignore the store operation for the final cost, and check that
+    // the resulting cost is > 0.
+    int Cost = 2 * VectorRegsNeeded + LastVectorHandling - 1;
+    assert(Cost > 0 && "Predicted cost of vector.reduce.add must be > 0");
+    return Cost;
+  }
```

Comment:
We are also supposed to use clang-format generally. In common code it's more or less obligatory, but in the backends not really - I usually follow the general coding style already used in that particular file I am editing. Sometimes I check with clang-format and decide to use / not use the suggested formatting - it may not always look as expected in the backend file comparing it to other code.

I tried in this case:

clang-format -i --lines="1250:1282" ./llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp

I think those two formatting fixes are improvements, so I would use them. I thought the std::min statement was a bit weird, and clang-format fixes it.

You can also do
clang/tools/clang-format/git-clang-format HEAD~1
to format the last commit (or more if you replace the '1')

Reply:
thanks, done :+1: